### ISSUE TITLE
Option to specify ClipboardItem is sensitive in .write() and .writeText()

### ISSUE THREAD

**Issue by sangeeth96** [w3c/clipboard-apis#154]
### Feature Request / Proposal

To add a boolean property `sensitive` to a `ClipboardItem`:

```js
const passwordBlob = new Blob(["secret"], { type: "text/plain" });

await navigator.clipboard.write([
    new ClipboardItem({
      [passwordBlob.type]: passwordBlob
      sensitive: true
    })
]);

// ...and maybe an options param to `.writeText()` as well?
await navigator.clipboard.writeText('secret', { sensitive: true });
```

### Why

Operating systems with built-in clipboard managers or browser's with clipboard managers (and maybe userland extensions?) could choose to not add such items to their history.

A real-world scenario that I've faced is while using 1Password on Windows 11, with Clipboard History enabled (<kbd>Win + V</kbd>). Using the native 1Password Windows app to copy a password/OTP doesn't persist it to Clipboard History. Not sure how they accomplish this.

But the same thing, when done via the 1Password extension, comes up in the Clipboard History. I'm not sure if the 1Password extension uses a different API that's available to extensions or utilize Clipboard API for this. Either way, being able to label items like this and if browsers handle this case appropriately, Clipboard Managers (and similar applications) can operate safely and not be prone to leak sensitive info.

### Concerns

- Such items would be easy targets for an existing malicious code
- Some users might prefer to _still_ keep these items in their history. But that's up to the browsers to decide if they'll provide such an option.

**Comment by BoCupp-Microsoft** [comment]
@snianu is there anything that would stop us from using the [UWP APIs on Windows](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.datatransfer.clipboardcontentoptions.isallowedinhistory?view=winrt-22000#Windows_ApplicationModel_DataTransfer_ClipboardContentOptions_IsAllowedInHistory) to specify whether an item should be added to the clipboard history or not based on a `ClipboardItem` option like @sangeeth96 has proposed?

@whsieh there are also [MacOS content options](https://developer.apple.com/documentation/appkit/nspasteboard/contentsoptions) that control some pasteboard behavior like whether the clipboard content can be roamed to another device.  Would it make sense to let the web author opt out of that behavior using a `ClipboardItem` option?

**Comment by BoCupp-Microsoft** [comment]
Note that [on Windows](https://docs.microsoft.com/en-us/uwp/api/windows.applicationmodel.datatransfer.clipboardcontentoptions?view=winrt-22000) it looks like roaming and history entries can be controlled per format.  I imagine its a performance feature so that large formats don't need to be realized and stored or transmitted over the network.

**Comment by mbrodesser** [comment]
Cc @mozfreddyb to get his view about the concerns mentioned in https://github.com/w3c/clipboard-apis/issues/154#issue-975114663.

**Comment by mozfreddyb** [comment]
I have a few clarifying questions.

If this is solely about providing this `sensitive` flag to the underlying operating system, so that other applications (e.g., cloud sync) can choose not to store them, then I think it's not a bad idea. Firefox is already setting this `sensitive` flag on certain platforms when copying passwords from the built-in password manager.

Do I also correctly understand that the browser wouldn't behave differently when *reading* a sensitive ClipboardItems, i.e., web sites would be able to have the same discretion for sensitive items?

It seems this whole issue is orthogonal to a different problem that's worth solving, which I want to explain for a second: An evil web page that is trying to paste indefinitely in the hope of scraping some passwords from the clipboard. It would be nice if we could agree that browser shouldn't web expose sensitive items through the read API without additional user interaction..

**Comment by mbrodesser** [comment]
Amendment to @mozfreddyb's post above: https://www.mozilla.org/en-US/security/advisories/mfsa2021-48/#CVE-2021-38505

**Comment by BoCupp-Microsoft** [comment]
> If this is solely about providing this sensitive flag to the underlying operating system, so that other applications (e.g., cloud sync) can choose not to store them, then I think it's not a bad idea.

Yes I believe your understanding is correct.

> It would be nice if we could agree that browser shouldn't web expose sensitive items through the read API without additional user interaction.

Do platform APIs exist to let app developers know if the item has been marked as sensitive?  I didn't spot any for Windows.

**Comment by annevk** [comment]
I'm not sure about the shape of this API. Is the problem here that this particular cloud implementation doesn't do end-to-end encryption? It seems a lot safer to assume that all data is sensitive unless the application explicitly indicates otherwise. Opting everyone into the "not sensitive" bucket seems bad? (It's also not entirely clear to me how an application would go about determining whether something is sensitive.)

**Comment by mbrodesser** [comment]
Given the different values (`ExcludeClipboardContentFromMonitorProcessing`, `CanIncludeInClipboardHistory`, `CanUploadToCloudClipboard`) of https://docs.microsoft.com/en-us/windows/win32/dataxchg/clipboard-formats#cloud-clipboard-and-clipboard-history-formats I wonder if the same level of granularity would make sense for `clipbboard.write()` and `clipboard.writeText()`.

**Comment by sangeeth96** [comment]
> If this is solely about providing this `sensitive` flag to the underlying operating system, so that other applications (e.g., cloud sync) can choose not to store them, then I think it's not a bad idea.

Yes, this was my intention. Passwords, API keys and anything that I as a developer firmly believe shouldn't go into a "Clipboard History" application, OS-provided or built-in can respect such a flag. Or, the browser can (if possible), signal this intention. Ideally, would prefer any such application could choose this for themselves. Plus, it could be a configurable option for these Clipboard managers whether to keep sensitive info in history or not. It's useful sometimes to copy these across devices but I understand that could pose a security risk too.


> Do I also correctly understand that the browser wouldn't behave differently when *reading* a sensitive ClipboardItems, i.e., web sites would be able to have the same discretion for sensitive items?

This is interesting. It could be helpful maybe to prevent accidentally sending a sensitive information in say a public forum. Websites could alert the user about it. Not entirely sure about the security implications besides a bad actor finding it easy to filter for these specific items.


> ...An evil web page that is trying to paste indefinitely in the hope of scraping some passwords from the clipboard.

Paste events don't need a user action to work? 🤔

**Comment by sangeeth96** [comment]
@annevk I totally agree with the "encrypt everything" point. But, if I take the example of Handoff and similar cross-device feature offered by other Clipboard management tools, I find it very convenient to copy passwords or OTPs like that. So, if I could choose as a user whether this should be allowed or not, that'd be awesome. Plus, if I'm screensharing and want to copy from my Clipboard manager a password I copied previously, having such a flag (and allowing copying of these) could let the Clipboard manager mask the password with *.

So, I personally think there's value in having something like this as a user and a developer.

**Comment by mbrodesser** [comment]
@sangeeth96 
> Paste events don't need a user action to work? :thinking: 

They need. That's also covered by the spec: https://w3c.github.io/clipboard-apis/#dom-clipboard-write. However, browsers currently don't behave exactly the same. Perhaps that might be clarified in https://github.com/w3c/clipboard-apis/pull/158; see also https://github.com/w3c/clipboard-apis/pull/158#discussion_r769627867.

**Comment by BoCupp-Microsoft** [comment]
> So, if I could choose as a user whether this should be allowed or not, that'd be awesome.

That's a good idea.  I could imagine a setting that would cause the UA to ignore these hints from the author.  I could also imagine UX that allows the user to override any author supplied hints at the time of copy.

Also, I think there are two hints that should be allowed separately as opposed to just one "sensitive" hint:

* history - controls which formats are saved
* roam - controls which formats can be shared across devices

You might not want to save a OTP in history, but you may want it shared to another of your devices.

**Comment by oliverdunk** [comment]
Just to share another example here, there is a community-driven standard set of types used on macOS by many applications: http://nspasteboard.org/

**Comment by snianu** [comment]
@sanketj FYI.

**Comment by dillonstreator** [comment]
Another use case worth considering: web apps that handle regulated data (healthcare, legal, finance).

The discussion so far has centered on passwords and OTPs, but there's a whole class of applications where a developer-supplied sensitivity hint would be a useful privacy control. Think PHI in healthcare apps (HIPAA), privileged material in legal tools, or special-category PII under GDPR.

Most of these applications are web-based now (EHRs, case management systems, patient portals) and copy/paste is a core workflow. A clinician copying a medical record number or lab result between systems is an everyday thing. On Windows with Cloud Clipboard enabled, that content gets uploaded to Microsoft's servers and synced across devices. Handoff does similar roaming on Apple platforms. For a covered entity, that's regulated data silently leaving the controlled environment.

Native apps can prevent this today (`CanUploadToCloudClipboard`, `ExcludeClipboardContentFromMonitorProcessing`, `org.nspasteboard.ConcealedType`) but web apps have no equivalent, which is a real platform gap for these industries.

This use case also speaks to a couple of earlier questions in the thread:

1. On "how would an application determine whether something is sensitive?": regulated applications aren't guessing. They have a legal obligation to classify this data and already do so internally. The hint just gives them a way to express a classification they already have.
2. It supports the per-capability design discussed above (history vs. roam) rather than a single boolean. Keeping PHI in local clipboard history may be acceptable in some deployments while cloud sync is not.

On user overrides: even if the UA ultimately lets users override the hint, it still has value for this class of app, since it gives enterprise or UA policy something concrete to enforce.
